### ISSUE TITLE
Getting SQL error code as zero with JDBC

### ISSUE THREAD

**Issue by debraj-manna** [cockroachdb/cockroach#19833]
I am getting SQL Error Code as zero with JDBC. As mentioned in this [discussion](https://forum.cockroachlabs.com/t/cockroach-transaction-throwing-exceptions/1071/10?u=tuk)  cockroach should not be returning zero error code.

The code to reproduce this-

```java
package com.db.cockroach;

import java.sql.*;


// Ambiguous whether the transaction committed or not.
class AmbiguousResultException extends SQLException{}

// A simple interface that provides a retryable lambda expression.
interface RetryableTransaction {
	public void run(Connection conn)
			throws SQLException, AmbiguousResultException;
}

public class App {
	public static RetryableTransaction transferFunds(final int id, final String name) {
		return new RetryableTransaction() {
			public void run(Connection conn)
					throws SQLException, AmbiguousResultException {
				// Check the current balance.
				conn.createStatement().executeUpdate(String.format("Insert into name_table values (%d,'%s')", id, name));
			}
		};
	}

	public static void retryTransaction(Connection conn, RetryableTransaction tx)
			throws SQLException, AmbiguousResultException {
		Savepoint sp = conn.setSavepoint("cockroach_restart");
		while(true) {
			try {
				// Attempt the transaction.
				tx.run(conn);
				conn.releaseSavepoint(sp);
				break;
			} catch(SQLException e) {
				// Check if the error code indicates a SERIALIZATION_FAILURE.
				if(e.getErrorCode() == 40001) {
					// Signal the database that we will attempt a retry.
					conn.rollback(sp);
				} else if(e.getErrorCode() == 40007) {
					throw new AmbiguousResultException();
				} else {
					// This is a not a serialization failure, pass it up the chain.
					throw e;
				}
			}
		}
		// Commit the transaction.
		conn.commit();
	}

	public static void main(String[] args) throws ClassNotFoundException, SQLException {
		// Load the postgres JDBC driver.
		Class.forName("org.postgresql.Driver");
		int id = 1;
		String name = "tuk";

		Connection db = DriverManager.getConnection("jdbc:postgresql://127.0.0.1:26257/vnera?sslmode=disable", "root", "");
		try {
			db.createStatement().executeUpdate(String.format("Insert into name_table values (%d,'%s')", id, name));
			db.setAutoCommit(false);
			RetryableTransaction transfer = transferFunds(id, name);
			retryTransaction(db, transfer);
		}catch(AmbiguousResultException e) {
			System.out.println("Ambiguous result encountered: " +e);
		} catch(SQLException e) {
			System.out.printf("SQLException encountered: %d, \n%s", e.getErrorCode(), e);
		}  finally {
			// Close the database connection.
			db.close();
		}
	}
}
```

OUTPUT-

```
SQLException encountered: 0, 
org.postgresql.util.PSQLException: ERROR: duplicate key value (id)=(1) violates unique constraint "primary"
```
The SQL schema for the table -

```
CREATE TABLE name_table (
    id integer NOT NULL,
    name text NOT NULL,
    PRIMARY KEY (id) 
);
```
Cockroach Version

```
dmanna-m01:main dmanna$ cockroach version
Build Tag:    v1.1.1
Build Time:   2017/10/19 15:29:38
Distribution: CCL
Platform:     darwin amd64
Go Version:   go1.8.3
C Compiler:   4.2.1 Compatible Clang 3.8.0 (tags/RELEASE_380/final)
Build SHA-1:  f3bc578c89627d02f2ca30f02b835388278a1c83
Build Type:   release
```

Postgres JAR

```
<dependency>
    <groupId>postgresql</groupId>
    <artifactId>postgresql</artifactId>
    <version>9.1-901-1.jdbc4</version>
</dependency>
```

**Comment by bdarnell** [comment]
The postgres JDBC driver returns error codes in `getSQLState()`, not `getErrorCode()`. This is because [not all postgres error codes are numeric](https://www.postgresql.org/docs/8.4/static/errcodes-appendix.html). See this [thread on the postgres mailing list](http://www.postgresql-archive.org/SQLException-getErrorCode-td2175792.html). 

Our java sample code appears to be incorrect and needs to be updated with this change.

**Comment by arjunravinarayan** [comment]
This was fixed in cockroachdb/docs#2126.
